### PR TITLE
Allow for slightly different content for different deployments

### PR DESCRIPTION
**Pull request by tessereth**
There are some differences we need between our deployment and 18F. This PR is tackling the content differences. There are a few different ways this could be done, including just changing the content. But I wanted to explore an approach where 18F and DTA could use the same code base. This PR does not change everything we need to update but provides a proof of concept that changes:

* some static text (eg the disclaimer)
* some links (eg CLI docs)
* The links in the top nav
* The content of the home page

It works by creating config files for each "skin" and aliasing the `skin` module in the webpack config based on the `$CG_SKIN` environment variable. This seems to work but given this is the first javascript codebase I've worked with (let alone react/webpack etc), I don't know if that's a good idea or not. Also, naming things is hard so some of the names could use a little work.

Most of it was pretty straight forward but the home page was a little more interesting. The home page seemed to be a collection of blocks, some of which applied to DTA and some of which didn't. Because the blocks include markup (lists, links etc) it seemed odd to put them in the config file. Instead I split them into components and put a list of components in the config file and dynamically put them in the grid. I did have to change the markup of the home page slightly for this to work.

So I have a few questions for 18F folk about this:

1. Does this seem like a sensible way of doing this?
2. Is this something you could see merging in upstream? I was thinking it might even be useful for you to be able to change a few things between the eastwest and govcloud deployments (eg the link to govcloud on the home page).
3. Anything else you feel like mentioning.

Thanks for your help!

### CODE REVIEW

**Comment by adborden** [comment]
@msecret would you take a look?

Awesome! The overall approach is pretty flexible, though eventually I would like to move to something more transparent. For example, each jsx component that allows skinning needs to be modified for the specific skin attributes. In the future, I'd like to see something like:

```
function findTemplate (templateName) {
  // Look for templateName component in CG_SKIN,
  // if not found, fall back to default templateName component
}
```

I'm not familiar with JSX enough to know how to plug-in to the "template" engine.

I would be okay merging something like this, though I'd like to take a more detailed look first. I think the current approach is okay for now, and I'd rather be including the DTA code sooner before we diverge too far. This does bring on some tech debt for us though, I can see the components becoming littered with `if (skinConfig.property)` conditionals if we stick with this approach for long. @tessereth do you have an idea of how heavy the customization will be on the DTA side, like a high-level roadmap of what problems need to be solved?

File: static_src/components/grid.jsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      rowTiles = [];
+      for (let j = i; j < i + cols && j < tiles.length; j++) {
+        Tile = tiles[j];
+        rowTiles.push(<Tile className={ this.styler(gridStyle) } key={j} />);
+      }
+      gridTiles.push(
+        <div className={ this.styler('grid') } key={i} >
+          {rowTiles.slice()}
+        </div>
+      );
```
Comment (adborden) [inline static_src/components/grid.jsx]: These for-loops are a bit hard to follow, I think you can write this using `map/reduce` to be a little more readable. Something like
```javascript
const rowSize = 12 / cols;
// 1. map tiles to html
// 2. group tiles into rows of rowSize
// 3. map rows to html
return tiles.map(function () {
  return <Tile className={ this.styler(gridStyle) } key={j} />;
}).reduce(function (memo, tile, index) {
  let rowIndex = Math.floor(index / rowSize);

  // Get/create the current row
  let row = memo[rowIndex] || [];

  // Add tile to the row
  row.push(tile);

  // Set the row to the grid
  memo[rowIndex] = row;
  return memo;
}, []).map(function (row) {
  return (<div className={ this.styler('grid') } key={i} >
    { row.slice() }
  </div>);
});
```

What do you think? Is this feedback too detailed for what you're looking for?

File: static_src/skins/cg/skin_config.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    ]
+  },
+  docs: {
+    cli: 'https://docs.cloud.gov/getting-started/setup/'
+  },
+  github: {
+    url: 'https://github.com/18F/cg-dashboard'
+  },
+  home: {
+    tiles: [HomeIntro, HomeCGStructure, HomeEnvironments, HomeSandbox, HomeAbout]
```
Comment (msecret) [inline static_src/skins/cg/skin_config.js]: So in our new design, we're actually getting to get rid of all of these in favor of showing quick org details like this:

screen shot 2016-11-28 at 4 46 32 pm

If that's the case, should there still be some sort of tile configuration?

File: static_src/components/disclaimer.jsx [hunk 3]
```diff
@@ -4,6 +4,7 @@ import style from 'cloudgov-style/css/cloudgov-style.css';
 import React from 'react';
 
 import createStyler from '../util/create_styler';
+import { skinConfig } from 'skin/skin_config.js';
```
Comment (msecret) [inline static_src/components/disclaimer.jsx]: I'd almost want to just call this `config`.

File: static_src/components/home.jsx [hunk 4]
```diff
@@ -15,42 +18,10 @@ export default class Home extends React.Component {
       <div>
         <div className={ this.styler('grid') }>
           <h1>Overview</h1>
-          <p className={ this.styler('page-dek') }>The dashboard provides an overview of your organizations, spaces, and applications. For full management and deployment of applications, use the <a href="https://docs.cloud.gov/getting-started/setup/">command line interface</a>.
+          <p className={ this.styler('page-dek') }>The dashboard provides an overview of your organizations, spaces, and applications. For full management and deployment of applications, use the <a href={ skinConfig.docs.cli }>command line interface</a>.
```
Comment (msecret) [inline static_src/components/home.jsx]: While I think this solution for docs links works for now, I do think there's a better solution for them. I'll starting thinking about it. That being said, we might not have much need for a more robust solution, maybe there won't be that many docs links.

File: static_src/skins/aus/skin_config.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import HomeAbout from '../../components/home_about.jsx';
```
Comment (msecret) [inline static_src/skins/aus/skin_config.js]: How would the skin files be used? So if another gov wanted to use this project, would they have to commit another skin file into the project? Would this `aus` skin file be part of the main `cg-dashboard` repo?

Comment (msecret) [inline static_src/skins/aus/skin_config.js]: I was thinking config would not be checked in, but I do agree it should be a js file, as otherwise it's too hard to do really useful things like bring in JSX components.

File: static_src/components/disclaimer.jsx (hunk 3 above)
Comment (tessereth) [inline static_src/components/disclaimer.jsx]: Cool. Learning new things!

Will update

**Comment by tessereth** [comment]
Thanks so much for your feedback!

> I can see the components becoming littered with if (skinConfig.property) conditionals if we stick with this approach for long.

I agree. I was trying to come up with something to avoid too many conditionals but it's still not perfect.

As for your questions about how much will be configured, at this point I think it's only the obviously cloud.gov stuff. So the links to your specific documentation and status pages, specific references to your deployments (eastwest/govcloud) that sort of thing. The only reason I did the tiling thing on the home page was to hide the stuff that didn't apply to us. If you change the homepage layout to general org info, then we won't need to configure it (so long as you don't hardcode org names or anything like that).

PS. Sorry for the late replies but I'm technically off work sick at the moment and probably will be for another couple of days. I'll have another closer look at this when I'm back in the office.

**Comment by adborden** [comment]
@tessereth speaking of updating the home/overview page, this is landing soon https://github.com/18F/cg-dashboard/pull/774

**Comment by tessereth** [comment]
Now updated for comments. Changes:
* `import {skinConfig} from 'skin/skin_config.js'` has changed to `import {config} from 'skin'`
* I reverted changes to the home page because your new version looks great and we can use it as-is
* The aus config now extends cg config so things can be added to the base config without breaking the other skins

My thoughts on how this process would work is:
* `skin/cg/index.js` would be checked into the 18F repo
* Our fork would add `skin/aus/index.js`
* If anyone else wants their own skin, they can fork and add their own or they can just clone and use a non-version-controlled file
* When anyone implements a feature, evaluate whether it has instance-specific things or just cloud-foundry-in-general things. If it's not obvious, just ask.
* Any instance-specific things should be made configurable with entries in `skin/cg/index.js`
* When anyone downstream (ie aus) pulls in changes, they should check `skin/cg/index.js` for any new entries that need overriding

Some principals for the config files themselves:
* All skins should "inherit" from the base config so entries can be added to the base without breaking other builds
* Simple content like links and plain text should be included in the config file itself
* Complex content requiring HTML markup should be implemented with different components for each skin, with the config file specifying which component(s)
* Complex content should be heavily discouraged and only used as a last resort

Thoughts?

**Comment by msecret** [comment]
@tessereth That process makes sense, I'm down to go ahead. I think we should try and  merge this into dashboard soon so any  new features we build will incorporate it. I can help out with this if needed.

This is a great start. If we find that the code is getting a little messy  with "if" statements,  we'll figure out something down the road.

Let me know what I can  do to help to get this merged into cg-dashboard.

**Comment by msecret** [comment]
Hey @tessereth. I've discussed with some people at 18F, and people are unsure of the interface to Immutable JS (the `getIn` calls for example), and felt simple objects would be easier to understand. I made the changes in https://github.com/AusDTO/cg-dashboard/pull/3 of how that would look. What do you think? We could probably be convinced that Immutable is the way to go if you have strong feelings.

**Comment by tessereth** [comment]
This PR has been replaced with 18F/cg-dashboard#793 and #5.
